ddl/index.go

```
@@ -846,48 +891,127 @@ func doReorgWorkForCreateIndex(w *worker, d *ddlCtx, t *meta.Meta, job *model.Jo
 		err = tryFallbackToTxnMerge(job, err)
 		return false, ver, errors.Trace(err)
 	}
-	if !done {
-		return false, ver, nil
-	}
-	err = bc.FinishImport(indexInfo.ID, indexInfo.Unique, tbl)
-	if err != nil {
-		if common.ErrFoundDuplicateKeys.Equal(err) {
-			err = convertToKeyExistsErr(err, indexInfo, tbl.Meta())
+	switch checkpoint.State {
+	case model.CheckpointStateNone:
+		checkpoint.State = model.CheckpointStateWriting
+		err := progressCheckpoint(w, job, checkpoint)
+		if err != nil {
+			return false, ver, errors.Trace(err)
 		}
-		if kv.ErrKeyExists.Equal(err) {
-			logutil.BgLogger().Warn("[ddl] import index duplicate key, convert job to rollback", zap.String("job", job.String()), zap.Error(err))
-			ver, err = convertAddIdxJob2RollbackJob(d, t, job, tbl.Meta(), indexInfo, err)
-		} else {
-			logutil.BgLogger().Warn("[ddl] lightning import error", zap.Error(err))
+		fallthrough
+	case model.CheckpointStateWriting:
+		done, ver, err = runReorgJobAndHandleErr(w, d, t, job, tbl, indexInfo, false)
+		if err != nil {
+			failpoint.Inject("MockCheckpointFailWriting", func(_ failpoint.Value) {
+				// Skip unregister to mock TiDB is killed.
+				failpoint.Return(false, ver, err)
+			})
+			ingest.LitBackCtxMgr.Unregister(job.ID)
 			err = tryFallbackToTxnMerge(job, err)
+			return false, ver, errors.Trace(err)
 		}
-		ingest.LitBackCtxMgr.Unregister(job.ID)
-		return false, ver, errors.Trace(err)
+		if !done {
+			return false, ver, nil
+		}
+		_, err := bc.Flush(indexInfo.ID)
+		if err != nil {
+			return false, ver, errors.Trace(err)
+		}
+		checkpoint.State = model.CheckpointStateEngineClosed
+		err = progressCheckpoint(w, job, checkpoint)
+		if err != nil {
+			return false, ver, errors.Trace(err)
+		}
+		return false, ver, nil
+	case model.CheckpointStateEngineClosed:
+		err = bc.EngMgr.Resume(bc, job.ID, indexInfo.ID, job.SchemaName, job.TableName)
+		if err != nil {
+			ingest.LitBackCtxMgr.Unregister(job.ID)
+			err = tryFallbackToTxnMerge(job, err)
```

Assume the lightning engine should be resume successful, then can we remove tryFallbackToTxnMerge?

---

Do you think we should return an error?